File: lib/scraper/join_scraper.py
import time

from ..driver.support.driver_support import DriverSupport
from ..asset.directory import Directory
from ..asset.url import URL
from selenium.common.exceptions import StaleElementReferenceException
from ..talker import Talker


class JoinScraper:
    """Scraper object for scraping ODs"""

    def __init__(self, driver, opts, nav_info, file_filter=None, sleep=False):
        """Initializes Scraper object

        :param WebDriver driver: Selenium Webdriver
        :param Opts opts: Opts class
        :param NavInfo nav_info: NavInfo object
        :param generic.Generic file_filter: Filter object to use for files
        :param bool sleep: stops the program before scraping
        """
        self._driver = driver
        self._opts = opts
        self._dirs = []
        self._files = []
        self._file_filter = file_filter() if file_filter else None
        self.nav_info = nav_info
        self._sleep = sleep

    def scrape(self, elements, directory) -> list:
        """Begin Scraping OD

        :param list elements: list of elements scraped
        :param Directory directory: current Directory
        """

        if elements is None:
            elements = self.scrape_items()
            if not elements and self._opts.refresh:
                # Refresh the page and try again
                self._driver.refresh()
                elements = self.scrape_items()
        self._store_links(elements, directory)
        return elements

    def scrape_items(self) -> list:
        """Retrieve scraped elements"""
        if self._sleep:
            time.sleep(4)
        return DriverSupport.get_elements_all(self._driver, self._opts,
                                              self.nav_info.css_select,
                                              self.nav_info.wait_err_message)

    def _store_links(self, elements, directory) -> None:
        """ Appends the element link onto the current URL

        :param list elements: list of elements
        :param Directory directory: current parent folder
        :return:
        """
        try:
            for el in elements:
                link = el.get_attribute(self.nav_info.css_attr)
                self._join_links(directory, link)
        except StaleElementReferenceException:
            Talker.warning("Element cannot be found on current page!", new_line=True)

    def _join_links(self, directory, link) -> None:
        """ Filter & join links together to make a new URL

        :param directory: current Parent directory
        :param link: new link to join together
        :return:
        """
        if JoinScraper._is_home(link):
            return
        link = self.apply_file_filter(link)
        if URL.is_a_file(link):
            self._files.append(URL.joiner(directory.url, link))
        else:
            new_level = directory.depth_level + 1
            if new_level < self._opts.depth:
                url = URL.joiner(directory.url, link)
                url.dir_transform()
                new_dir = Directory(new_level, url)
                self._dirs.append(new_dir)

    @staticmethod
    def _all_filters(text) -> bool:
        return JoinScraper._is_home(text) or JoinScraper._is_js_void(text)

    @staticmethod
    def _is_home(text):
        home_list = {'/', '.', '..', '../', './'}
        init_total = len(home_list)
        home_list.add(text)
        return init_total == len(home_list)

    @staticmethod
    def _is_js_void(text):
        return "javascript:" in text

    def apply_file_filter(self, link):
        if self._file_filter:
            return self._file_filter.apply(link)
        return link

    def reset(self) -> None:
        """Reset the list fields"""
        self._files = []
        self._dirs = []

    def get_results(self) -> tuple[list, list]:
        """Retrieve the files/directories as a result of scraping

        :return: list of files and directories
        """
        return self._dirs, self._files

File: app/src/main/java/xyz/zimtools/zyod/scrapers/JoinScraper.java
package xyz.zimtools.zyod.scrapers;

import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.RemoteWebDriver;
import xyz.zimtools.zyod.AppConfig;
import xyz.zimtools.zyod.args.Args;
import xyz.zimtools.zyod.assets.Directory;
import xyz.zimtools.zyod.assets.ODUrl;
import xyz.zimtools.zyod.assets.info.NavInfo;
import xyz.zimtools.zyod.scrapers.filters.ScrapeFilter;
import xyz.zimtools.zyod.support.NavSupport;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Scrape ODs using the {@link WebElement}'s existing attributes.
 */
public class JoinScraper implements ODScraper {
    protected final RemoteWebDriver driver;
    protected final Args args;
    protected final ScrapeFilter filter;
    protected final NavInfo navInfo;
    protected List<ODUrl> files;
    protected List<Directory> dirs;

    public JoinScraper(RemoteWebDriver driver, Args args, NavInfo navInfo, ScrapeFilter filter) {
        this.driver = driver;
        this.args = args;
        this.navInfo = navInfo;
        this.filter = filter;
        this.files = new ArrayList<>(); // Holds all scraped file links
        this.dirs = new ArrayList<>(); // Holds all scraped directories
    }

    public JoinScraper(RemoteWebDriver driver, Args args, NavInfo navInfo) {
        this(driver, args, navInfo, null);
    }

    @Override
    public NavInfo getNavInfo() {
        return this.navInfo;
    }

    @Override
    public List<ODUrl> getFiles() {
        return this.files;
    }

    @Override
    public List<Directory> getDirs() {
        return this.dirs;
    }

    /**
     * Begins scraping the current OD
     *
     * @param elements  list of scraped {@link WebElement}s
     * @param parentDir the current parent {@link Directory} to scrape
     * @return list of scraped WebElements
     */
    @Override
    public List<WebElement> scrape(List<WebElement> elements, Directory parentDir) {
        if (elements.isEmpty()) {
            elements = this.scrapeItems();
            elements = this.refresh(elements);
        }
        this.storeLinks(elements, parentDir);
        return elements;
    }

    /**
     * Scrape for {@link WebElement}s using CSS file selector
     *
     * @return list of scraped WebElements
     */
    @Override
    public List<WebElement> scrapeItems() {
        long waitTime = this.args.getArgsScraper().getWait();
        if (waitTime > 0) {
            AppConfig.sleep(waitTime);
        }
        return NavSupport.getElements(this.driver, this.navInfo.getCssFileSelector());
    }


    /**
     * Appends the link found in {@link WebElement}s onto the current URL and save them.
     *
     * @param elements  list of scraped WebElements
     * @param parentDir current parent directory
     */
    private void storeLinks(List<WebElement> elements, Directory parentDir) {
        if (elements.isEmpty()) {
            return;
        }
        for (WebElement el : elements) {
            this.joinLoop(el, parentDir);
        }
    }

    /**
     * Join specified attribute item found in {@link WebElement} to the current URL.
     *
     * @param element   scraped WebElement
     * @param parentDir the current parent {@link Directory}
     */
    protected void joinLoop(WebElement element, Directory parentDir) {
        String fileLink = element.getAttribute(this.navInfo.getCssAttr());
        this.joinLinks(fileLink, parentDir);
    }

    /**
     * Filter & join links together to make a new URL.
     *
     * @param link      new link to join together
     * @param parentDir current parent {@link Directory}
     */
    protected void joinLinks(String link, Directory parentDir) {
        if (this.isHome(link) || this.isJSVoid(link)) {
            return;
        }
        // NOTE: Files with extension length > 7 & < 3 are considered directories.
        link = this.applyFilter(link);
        if (ODUrl.isFile(link)) {
            ODUrl joinedURL = ODUrl.joiner(parentDir.toString(), link);
            this.files.add(joinedURL);
        } else {
            int newDepthLevel = parentDir.getDepthLevel() + 1;
            ODUrl dirURL = ODUrl.joiner(parentDir.toString(), link);
            dirURL.dirTransform();
            Directory newDir = new Directory(newDepthLevel, dirURL);
            if (newDepthLevel < this.args.getArgsNavigator().getDepth()) {
                this.dirs.add(newDir);
            }
        }
    }

    /**
     * Refresh the page.
     * <p>
     * Current page will refresh in an attempt to retrieve {@link WebElement}s.
     * </p>
     *
     * @return list of WebElements on current page.
     */
    protected List<WebElement> refresh(List<WebElement> elements) {
        if (elements.isEmpty() && this.args.getArgsMisc().isRefreshing()) {
            this.driver.navigate().refresh();
            elements = this.scrapeItems();
        }
        return elements;
    }

    /**
     * Check if text points to a home directory
     * <p>
     * Determine if specified text leads to a previous parent directory, homepage, or front
     * page.
     * </p>
     *
     * @param text the text to check
     * @return true if text leads to a home directory or previous parent directory; false
     * otherwise.
     */
    private boolean isHome(String text) {
        String[] homeList = {"/", ".", "..", "../", "./"};
        return Arrays.asList(homeList).contains(text);
    }

    /**
     * Check if {@code javascript:} is present.
     * <p>
     * {@code javascript:} usually indicate the execution of a JavaScript code, which is not
     * needed when attempting to retrieve OD resources.
     * </p>
     *
     * @param text the text to check.
     * @return true if {@code javascript:} is present; otherwise false.
     */
    private boolean isJSVoid(String text) {
        return text.contains("javascript:");
    }

    /**
     * Apply Scraper filter.
     *
     * @param text the text to apply the filter to
     * @return the filter text
     */
    @Override
    public String applyFilter(String text) {
        return this.filter != null ? this.filter.apply(text) : text;
    }

    /**
     * Reset records of all files and directories scraped
     */
    @Override
    public void reset() {
        this.files.clear();
        this.dirs.clear();
    }
}